### PR TITLE
Add more TypeScript types

### PR DESCRIPTION
**Pull request by lfades**
The purpose of this PR is to add missing types that can be very useful for Typescript projects using Next.js and for the core too.

The types are not yet being published to NPM, that will come in another PR.

### CODE REVIEW

**Review by timneutkens — APPROVED** [review]

File: packages/next/pages/_app.tsx [hunk 1]
```diff
@@ -69,47 +75,47 @@ const warnUrl = execOnce(() => {
...
-    push: (url, as) => {
+    push: (url: string, as?: string) => {
       warnUrl()
       return router.push(url, as)
     },
-    pushTo: (href, as) => {
+    pushTo: (href: string, as?: string) => {
       warnUrl()
-      const pushRoute = as ? href : null
+      const pushRoute = as ? href : ''
```
Comment (Timer) [inline packages/next/pages/_app.tsx]: Why was this switched from `null`?

File: packages/next/pages/_app.tsx [hunk 2]
```diff
@@ -69,47 +75,47 @@ const warnUrl = execOnce(() => {
...
-    replace: (url, as) => {
+    replace: (url: string, as?: string) => {
       warnUrl()
       return router.replace(url, as)
     },
-    replaceTo: (href, as) => {
+    replaceTo: (href: string, as?: string) => {
       warnUrl()
-      const replaceRoute = as ? href : null
+      const replaceRoute = as ? href : ''
```
Comment (Timer) [inline packages/next/pages/_app.tsx]: Why was this switched from `null`?

**Review by Timer — APPROVED** [review]
Seems fine if these null changes were intentional.

File: packages/next/pages/_app.tsx (hunk 1 above)
Comment (lfades) [reply packages/next/pages/_app.tsx]: @Timer To keep it with the same type used by `href`, which is always a `string` or `undefined`

File: packages/next/pages/_app.tsx (hunk 2 above)
Comment (lfades) [reply packages/next/pages/_app.tsx]: @Timer To keep it with the same type used by href, which is always a string

**Review by ijjk — APPROVED** [review]

File: packages/next-server/lib/utils.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+export interface IDocumentContext extends IContext {
+  renderPage: RenderPage
+}
+
+export interface IDocumentInitialProps extends RenderPageResult {
+  styles?: React.ReactElement[]
+}
+
+export interface IDocumentProps extends IDocumentInitialProps {
+  __NEXT_DATA__: INEXTDATA
```
Comment (huv1k) [inline packages/next-server/lib/utils.ts]: What is reasoning to have it uppercase?

**Pull request merged**